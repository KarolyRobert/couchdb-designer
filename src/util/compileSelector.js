
const selectorTypes = {
    "$lt": "any",
    "$gt": "any",
    "$lte": "any",
    "$gte": "any",
    "$eq": "any",
    "$ne": "any",
    "$exists": "boolean",
    "$type": "type",
    "$in" : "array",
    "$nin" : "array",
    "$size" : "integer",
    "$mod" : "mod",
    "$regex": "regex",
    "$and" : "selectors",
    "$or": "selectors",
    "$not": "selector",
    "$nor": "selectors",
    "$all": "array",
    "$elemMatch": "selector",
    "$allMatch": "selector",
    "$keyMapMatch": "selector",
}

const getValidArgument = selectorKey => {
    if(selectorKey in selectorTypes) {
        return selectorTypes[selectorKey];
    }else{
        throw('invalid selector');
    }
}

const compileSelector = (selector,file,index, selectorKey = false) => {
    if(selectorKey) {
        let selectorType = selector === null ? 'null' : Array.isArray(selector) ? 'array' : typeof selector;
        if(selectorKey.startsWith('$')){
            let argumentType = getValidArgument(selectorKey);
            switch(argumentType){
                case 'any':
                    return selector;
                case 'boolean':
                case 'array':
                    if(selectorType === argumentType){
                        return selector;
                    }else{
                        throw(`${selectorKey}'s argument must be ${argumentType}! You can fix it in ${file} / ${index}`);
                    }
                case 'integer':
                    if(Number.isInteger(selector)){
                        return selector;
                    }else{
                        throw(`${selectorKey}'s argument must be ${argumentType}! You can fix it in ${file} / ${index}`);
                    }
                case 'regex':
                    if(selectorType === 'string'){
                        return selector;
                    }else{
                        throw(`${selectorKey}'s argument must be a regular expression! You can fix it in ${file} / ${index}`);
                    }
                case 'type':
                    if(['null','boolean','number','string','array','object'].includes(selector)){
                        return selector;
                    }else{
                        throw(`${selectorKey}'s argument must be one of the null,boolean,number,string,array,object values! You can fix it in ${file} / ${index}`);
                    }
                case 'mod':
                    if(selectorType === 'array' && selector.length === 2 && Number.isInteger(selector[0]) && Number.isInteger(selector[1])){
                        return selector;
                    }else{
                        throw('$mod argument must be an two element array of integers! You can fix it in ${file} / ${index}');
                    }
                case 'selector':
                    if(selectorType === 'object'){
                        return compileSelector(selector,file,index);
                    }else{
                        throw(`${selectorKey}'s argument must be ${argumentType}! You can fix it in ${file} / ${index}`);
                    }
                case 'selectors':
                    if(selectorType === 'array'){
                        let selectorArray = [];
                        for(let partSelector of selector){
                            try{
                            selectorArray.push(compileSelector(partSelector,file,index));
                            }catch(err){
                                throw(`${selectorKey}'s argument must be an array of selectors! You can fix it in ${file} / ${index}`);
                            }
                        }
                        return selectorArray;
                    }else{
                        throw(`${selectorKey}'s argument must be an array of selectors! You can fix it in ${file} / ${index}`);
                    }

            }
        }else{
            if(selectorType === 'array' || selectorType === 'string' || selectorType === 'number'|| selectorType === 'boolean' || selectorType === 'null'){
                return {"$eq":selector};
            }else{
                if(Object.keys(selector).length){
                    return compileSelector(selector,file,index);
                }else{
                    return {"$eq":selector};
                }
            }
        }
    }else if(typeof selector === 'object') {
        let compiledSelector = {};
        let keys = Object.keys(selector);
        if(keys.length > 1){
            let andSelector = []
            for(let key of keys){
                andSelector.push({[key]:compileSelector(selector[key],file,index,key)});
            }
            compiledSelector['$and'] = andSelector;
        }else if (keys.length === 1){
            compiledSelector[keys[0]] = compileSelector(selector[keys[0]],file,index,keys[0]);
        }
        return compiledSelector;
    }else{
        throw('The selector must be an object! You can fix it in ${file} / ${index}');
    }
}

export default compileSelector;