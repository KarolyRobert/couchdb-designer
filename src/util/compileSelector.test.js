import compileSelector from './compileSelector';

const fullSelector = {
    foo:'bar',
    bar:{
        foo:'bar',
        bar:{
            baz:{
                "$lte":'bar',
                "$gte":'foo',
            },
            foo:{
                "$exists":false
            },
            bar:{
                "$type":'array'
            }
        }
    },
    baz:{
        "$lt": 2010,
        "$gt": 1000,
        "$ne": 1500
    },
    foobar:{
        "$in":[1,2],
        "$nin":[3,4],
        "$size":5,
        "$all":[6,7],
        "$elemMatch":{"$eq":10},
        "$allMatch":{"$lt":20,"$gt":0}
    },
    barbaz:{
        "$mod":[11,3],
        "$regex":"^foo"
    },
    foobaz:{
        "$and":[
            {foo:'bar'},
            {bar:{
                "$or":[
                    {foo:'bar'},
                    {bar:'baz'},
                    {"$not":{baz:'foo'}},
                ]
            }}
        ]
    },
    bazbar:{
        "$keyMapMatch":{
            "$eq":'bar'
        }
    }
}

const error1 = {
    foo:{
        "$exists": "true"
    }
}

const error2 = {
    foo:{
        "$in": "array"
    }
}

const error3 = {
    foo:{
        "$size": 12.34
    }
}


const error4 = {
    foo:{
        "$regex": 12.34
    }
}


const error5 = {
    foo:{
        "$type": 'integer'
    }
}


const error6 = {
    foo:{
        "$mod": 'integer'
    }
}



const error7 = {
    foo:{
        "$not": 'integer'
    }
}



const error8 = {
    foo:{
        "$nor": ['integer']
    }
}


describe('compileSelector',() => {
    test('all case',() => {
        expect(compileSelector(fullSelector,'file','index')).toMatchSnapshot();
    })
    test('errors',() => {
        expect(() => compileSelector(error1,'file','index')).toThrow("$exists's argument must be boolean!");
        expect(() => compileSelector(error2,'file','index')).toThrow("$in's argument must be array!");
        expect(() => compileSelector(error3,'file','index')).toThrow("$size's argument must be integer!");
        expect(() => compileSelector(error4,'file','index')).toThrow("$regex's argument must be a regular expression!");
        expect(() => compileSelector(error5,'file','index')).toThrow("$type's argument must be one of the null,boolean,number,string,array,object values!");
        expect(() => compileSelector(error6,'file','index')).toThrow("$mod argument must be an two element array of integers!");
        expect(() => compileSelector(error7,'file','index')).toThrow("$not's argument must be selector!");
        expect(() => compileSelector(error8,'file','index')).toThrow("$nor's argument must be an array of selectors!");
    })
});