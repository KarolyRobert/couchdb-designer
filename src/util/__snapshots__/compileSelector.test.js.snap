// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compileSelector all case 1`] = `
Object {
  "$and": Array [
    Object {
      "foo": Object {
        "$eq": "bar",
      },
    },
    Object {
      "bar": Object {
        "$and": Array [
          Object {
            "foo": Object {
              "$eq": "bar",
            },
          },
          Object {
            "bar": Object {
              "$and": Array [
                Object {
                  "baz": Object {
                    "$and": Array [
                      Object {
                        "$lte": "bar",
                      },
                      Object {
                        "$gte": "foo",
                      },
                    ],
                  },
                },
                Object {
                  "foo": Object {
                    "$exists": false,
                  },
                },
                Object {
                  "bar": Object {
                    "$type": "array",
                  },
                },
              ],
            },
          },
        ],
      },
    },
    Object {
      "baz": Object {
        "$and": Array [
          Object {
            "$lt": 2010,
          },
          Object {
            "$gt": 1000,
          },
          Object {
            "$ne": 1500,
          },
        ],
      },
    },
    Object {
      "foobar": Object {
        "$and": Array [
          Object {
            "$in": Array [
              1,
              2,
            ],
          },
          Object {
            "$nin": Array [
              3,
              4,
            ],
          },
          Object {
            "$size": 5,
          },
          Object {
            "$all": Array [
              6,
              7,
            ],
          },
          Object {
            "$elemMatch": Object {
              "$eq": 10,
            },
          },
          Object {
            "$allMatch": Object {
              "$and": Array [
                Object {
                  "$lt": 20,
                },
                Object {
                  "$gt": 0,
                },
              ],
            },
          },
        ],
      },
    },
    Object {
      "barbaz": Object {
        "$and": Array [
          Object {
            "$mod": Array [
              11,
              3,
            ],
          },
          Object {
            "$regex": "^foo",
          },
        ],
      },
    },
    Object {
      "foobaz": Object {
        "$and": Array [
          Object {
            "foo": Object {
              "$eq": "bar",
            },
          },
          Object {
            "bar": Object {
              "$or": Array [
                Object {
                  "foo": Object {
                    "$eq": "bar",
                  },
                },
                Object {
                  "bar": Object {
                    "$eq": "baz",
                  },
                },
                Object {
                  "$not": Object {
                    "baz": Object {
                      "$eq": "foo",
                    },
                  },
                },
              ],
            },
          },
        ],
      },
    },
    Object {
      "bazbar": Object {
        "$keyMapMatch": Object {
          "$eq": "bar",
        },
      },
    },
  ],
}
`;
